### PR TITLE
WIP: MultisigControl Smart Contract Spec

### PR DESCRIPTION
**Pull request by C0deMunk33**
Create specification for MultisigControl smart contract and how that functions.

see Multisig_Control_Spec branch

### CODE REVIEW

**Comment by C0deMunk33** [comment]
This is now open for comment.

File: specs/0000-multisig_control_spec.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+Specification PR: https://gitlab.com/vega-protocol/product/merge_requests
+
+
+# Summary
+“Whoever proposes an action aggregates and submits signatures and thus pays gas” - Danny
+
+In order for the Vega network to authorize function execution on Ethereum smart contracts, a mechanism needs to be created to verify that this is the will of Vega while placing the burden of execution costs on the proposer of the function execution. To do this, we have created a multisig process that enables a proposer to aggregate and submit a number of Vega validator node signatures in order to execute any Vega controlled smart contract function.       
+
+# Guide-level explanation
+Vega controls and maintains a number of Ethereum smart contracts which have functions that can only be ran once authorized by Vega consensus and are always requested by an interested party.
```
Comment (ashleyvega) [inline specs/0000-multisig_control_spec.md]: can only be *run*

File: specs/0000-multisig_control_spec.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+# Guide-level explanation
+Vega controls and maintains a number of Ethereum smart contracts which have functions that can only be ran once authorized by Vega consensus and are always requested by an interested party. 
+For instance, for depositing of settlement instrument assets, such as Ether or DAI, Vega has launched a number of "bridge" smart contracts. These contracts contain functions that are controlled and thus only authorizable from Vega consensus. These functions include withdrawing and whitelisting assets.
+
+As an example: once a user has requested a withdrawal and Vega consensus has agreed that the withdrawal should happen, the user will be presented with a number of signed orders from validator nodes that exceeds the threshold of signatures required. The user will then submit this "signature bundle" to the withdrawal function on the bridge smart contract along with the asset type and amount.
```
Comment (ashleyvega) [inline specs/0000-multisig_control_spec.md]: I assume the signatures cover the asset type and amount, but do the signatures cover any other data that needs to be mentioned?

File: specs/0000-multisig_control_spec.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    //returns true if nonce has been used
+    function is_nonce_used(uint nonce) public view returns(bool);
+}
+```  
+
+# Acceptance Criteria
+### MultisigControl Smart Contract 
+* MultisigControl smart contract is deployed to Ethereum testnet (Ropsten)
+* Set Threshold
+  * A valid signature bundle, threshold (in hundredths of %), and unused nonce can be passed to `set_threshold` function to set the approval threshold in hundredths of a percent (`TODO: check this mechanism/math`)
```
Comment (ashleyvega) [inline specs/0000-multisig_control_spec.md]: tenths of a percent (cf line 70) ?

File: specs/0000-multisig_control_spec.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  * A successful call to `set_threshold` emits `ThresholdSet` event
+  * Subsequent calls to `get_current_threshold()` returns updated threshold value
+  * An invalid signature passed to `set_threshold` function is rejected 
+  * A threshold passed to `set_threshold` outside of sane range is rejected
+  * A nonce passed to `set_threshold` that has already been used is rejected
+* Add Signer 
+  * A valid signature bundle, non-signer ethereum address, and unused nonce can be passed to `add_signer` function to add an Ethereum address of a new signer to the list of approved signers
```
Comment (ashleyvega) [inline specs/0000-multisig_control_spec.md]: [n00b smart contract question alert] How do we store the list of used nonces?

**Review by ashleyvega — COMMENTED** [review]
Just a few small comments/questions.

File: specs/0000-multisig_control_spec.md (hunk 2 above)
Comment (C0deMunk33) [reply specs/0000-multisig_control_spec.md]: If you look at elsewhere in this spec it shows the pattern

File: specs/0000-multisig_control_spec.md (hunk 3 above)
Comment (C0deMunk33) [reply specs/0000-multisig_control_spec.md]: Yeah, you are correct... I had a TODO around this being logical

File: specs/0000-multisig_control_spec.md (hunk 4 above)
Comment (C0deMunk33) [reply specs/0000-multisig_control_spec.md]: it's stored in a `mapping(uint256=>bool)`

File: specs/0000-multisig_control_spec.md (hunk 2 above)
Comment (C0deMunk33) [reply specs/0000-multisig_control_spec.md]: 3rd paragraph under `Signature Bundles` to be specific

File: specs/0000-multisig_control_spec.md (hunk 4 above)
Comment (ashleyvega) [reply specs/0000-multisig_control_spec.md]: :+1: thanks.

File: specs/0000-multisig_control_spec.md (hunk 2 above)
Comment (ashleyvega) [reply specs/0000-multisig_control_spec.md]: Ah, the `encode(encode(many things), validating_contract_or_submitter_address)`.

**Comment by edd** [comment]
@Vegaklaus @barnabee Please review

**Review by barnabee — APPROVED** [review]
I can't comment on security matters (needs strong internal review, testing and multiple external audits before it gets to mainnet...) but this looks good to me

**Comment by C0deMunk33** [comment]
@edd, there's a file 0000-price-monitoring.md that should not be here, does anyone know what that is about?

Other than that, this is complete, would you (@edd ) please assign 0000-multisig_control_spec.md a number and merge at your convenience?

**Pull request merged**